File: backend-python/app/modules/system/service/role_service.py
from __future__ import annotations

from sqlalchemy import String, delete, func, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.common.errors import BusinessError, bad_request, not_found
from app.common.id import new_id
from app.common.pagination import PageQuery, PageResult
from app.modules.system.dto.role_request import MenuAssignRequest, RoleSaveRequest
from app.modules.system.entity import SysMenu, SysRole, SysRoleMenu, SysUser, SysUserRole


class RoleService:
    def __init__(self, db: AsyncSession) -> None:
        self.db = db

    async def list_roles(
        self,
        page: PageQuery,
        role_name: str | None,
        role_code: str | None,
        status: int | None,
        created_at: str | None,
    ) -> PageResult[dict]:
        filters = [SysRole.deleted == 0]
        if role_name and role_name.strip():
            filters.append(SysRole.role_name.like(f"%{role_name.strip()}%"))
        if role_code and role_code.strip():
            filters.append(SysRole.role_code.like(f"%{role_code.strip()}%"))
        if status is not None:
            filters.append(SysRole.status == status)
        if created_at and created_at.strip():
            filters.append(SysRole.created_at.cast(String).like(f"%{created_at.strip()}%"))
        total = await self.db.scalar(
            select(func.count()).select_from(SysRole).where(*filters)
        ) or 0
        rows = (
            await self.db.scalars(
                select(SysRole)
                .where(*filters)
                .order_by(SysRole.created_at.desc(), SysRole.id.desc())
                .offset((page.page - 1) * page.page_size)
                .limit(page.page_size)
            )
        ).all()
        return PageResult(
            list=[self._vo(row) for row in rows],
            total=str(total),
            page=page.page,
            pageSize=page.page_size,
        )

    async def detail(self, role_id: int) -> dict:
        return self._vo(await self._find(role_id))

    async def options(self) -> list[dict]:
        rows = (
            await self.db.scalars(
                select(SysRole).where(SysRole.deleted == 0).order_by(SysRole.id)
            )
        ).all()
        return [self._vo(row) for row in rows]

    async def permissions(self, role_id: int) -> dict:
        await self._find(role_id)
        menu_ids = (
            await self.db.scalars(
                select(SysRoleMenu.menu_id)
                .where(SysRoleMenu.role_id == role_id)
                .order_by(SysRoleMenu.menu_id)
            )
        ).all()
        return {
            "menuIds": [str(menu_id) for menu_id in menu_ids],
            "permissionCodes": [],
        }

    async def create(self, request: RoleSaveRequest) -> str:
        await self._assert_role_code_available(request.role_code.strip(), None)
        row = SysRole(
            id=new_id(),
            role_name=request.role_name.strip(),
            role_code=request.role_code.strip(),
            builtin=0,
            status=1 if request.status is None else request.status,
            remark=_optional(request.remark),
            deleted=0,
        )
        self.db.add(row)
        await self.db.commit()
        return str(row.id)

    async def update(self, role_id: int, request: RoleSaveRequest) -> None:
        row = await self._find(role_id)
        await self._assert_role_code_available(request.role_code.strip(), role_id)
        row.role_name = request.role_name.strip()
        row.role_code = request.role_code.strip()
        row.status = 1 if request.status is None else request.status
        row.remark = _optional(request.remark)
        await self.db.commit()

    async def delete(self, role_id: int) -> None:
        row = await self._find(role_id)
        if row.builtin == 1:
            raise bad_request("operation failed")
        count = await self.db.scalar(
            select(func.count()).select_from(SysUserRole).where(SysUserRole.role_id == role_id)
        )
        if count:
            raise BusinessError(409000, "operation failed")
        row.deleted = 1
        await self.db.commit()

    async def update_status(self, role_id: int, status: int | None) -> None:
        row = await self._find(role_id)
        row.status = 1 if status is None else status
        await self.db.commit()

    async def assign_menus(self, role_id: int, request: MenuAssignRequest) -> None:
        await self._find(role_id)
        if len(request.menu_ids) != len(set(request.menu_ids)):
            raise bad_request("operation failed")
        if request.menu_ids:
            count = await self.db.scalar(
                select(func.count()).select_from(SysMenu).where(
                    SysMenu.id.in_(request.menu_ids),
                    SysMenu.deleted == 0,
                )
            )
            if count != len(request.menu_ids):
                raise bad_request("operation failed")
        await self.db.execute(delete(SysRoleMenu).where(SysRoleMenu.role_id == role_id))
        self.db.add_all(
            SysRoleMenu(id=new_id(), role_id=role_id, menu_id=menu_id)
            for menu_id in request.menu_ids
        )
        await self.db.commit()

    async def users(self, role_id: int, page: PageQuery) -> PageResult[dict]:
        await self._find(role_id)
        filters = [
            SysUser.deleted == 0,
            SysUser.id.in_(
                select(SysUserRole.user_id).where(SysUserRole.role_id == role_id)
            ),
        ]
        total = await self.db.scalar(
            select(func.count()).select_from(SysUser).where(*filters)
        ) or 0
        rows = (
            await self.db.scalars(
                select(SysUser)
                .where(*filters)
                .order_by(SysUser.created_at.desc(), SysUser.id.desc())
                .offset((page.page - 1) * page.page_size)
                .limit(page.page_size)
            )
        ).all()
        return PageResult(
            list=[self._user_vo(row) for row in rows],
            total=str(total),
            page=page.page,
            pageSize=page.page_size,
        )

    async def _find(self, role_id: int) -> SysRole:
        row = await self.db.scalar(
            select(SysRole).where(SysRole.id == role_id, SysRole.deleted == 0)
        )
        if row is None:
            raise not_found()
        return row

    async def _assert_role_code_available(self, role_code: str, role_id: int | None) -> None:
        query = select(SysRole.id).where(
            SysRole.role_code == role_code,
            SysRole.deleted == 0,
        )
        if role_id is not None:
            query = query.where(SysRole.id != role_id)
        if await self.db.scalar(query) is not None:
            raise BusinessError(409000, "roleCode already exists")

    @staticmethod
    def _vo(row: SysRole) -> dict:
        return {
            "id": str(row.id),
            "roleName": row.role_name,
            "roleCode": row.role_code,
            "builtin": row.builtin,
            "status": row.status,
            "remark": row.remark,
            "createdAt": row.created_at,
            "updatedAt": row.updated_at,
        }

    @staticmethod
    def _user_vo(row: SysUser) -> dict:
        return {
            "id": str(row.id),
            "username": row.username,
            "realName": row.real_name,
            "phone": row.phone,
            "email": row.email,
            "avatar": row.avatar,
            "status": row.status,
            "deptId": str(row.dept_id) if row.dept_id is not None else None,
            "remark": row.remark,
            "lastLoginAt": row.last_login_at,
            "createdAt": row.created_at,
            "updatedAt": row.updated_at,
        }


def _optional(value: str | None) -> str | None:
    return value.strip() if value and value.strip() else None

File: backend-rust/src/modules/system/service/role_service.rs
use crate::common::{AppError, I64String, PageResult, next_id};
use crate::modules::system::dto::role_request::{RoleQuery, RoleSaveRequest};
use crate::modules::system::entity::sys_role::SysRole;
use crate::modules::system::entity::sys_user::SysUser;
use crate::modules::system::vo::role_permission_vo::RolePermissionVo;
use rbatis::RBatis;
use serde::Deserialize;
use std::collections::BTreeSet;
use std::sync::Arc;

#[derive(Debug, Deserialize)]
struct CountRow {
    total: i64,
}

#[derive(Debug, Deserialize)]
struct IdRow {
    id: I64String,
}

pub async fn list(
    database: Option<&Arc<RBatis>>,
    query: &RoleQuery,
) -> Result<PageResult<SysRole>, AppError> {
    let database = require_database(database)?;
    let params = query.normalize_page()?;
    let offset = ((params.page - 1) * params.page_size) as i64;
    let (where_sql, args) = role_where(query);

    let mut page_args = args.clone();
    page_args.push(rbs::value!(params.page_size));
    page_args.push(rbs::value!(offset));
    let rows: Vec<SysRole> = database
        .exec_decode(
            &format!(
                "select id, role_name, role_code, builtin, status, remark, deleted, created_at, updated_at from sys_role {where_sql} order by created_at desc limit ? offset ?"
            ),
            page_args,
        )
        .await
        .map_err(map_database_error)?;
    let total = count(
        database,
        &format!("select count(*) as total from sys_role {where_sql}"),
        args,
    )
    .await?;
    Ok(PageResult {
        list: rows,
        total: I64String(total),
        page: params.page,
        page_size: params.page_size,
    })
}

pub async fn options(database: Option<&Arc<RBatis>>) -> Result<Vec<SysRole>, AppError> {
    let database = require_database(database)?;
    database
        .exec_decode(
            "select id, role_name, role_code, builtin, status, remark, deleted, created_at, updated_at from sys_role where deleted = 0 order by created_at desc",
            vec![],
        )
        .await
        .map_err(map_database_error)
}

pub async fn detail(database: Option<&Arc<RBatis>>, id: i64) -> Result<SysRole, AppError> {
    let database = require_database(database)?;
    detail_with_database(database, id).await
}

pub async fn users(
    database: Option<&Arc<RBatis>>,
    role_id: i64,
    query: &RoleQuery,
) -> Result<PageResult<SysUser>, AppError> {
    let database = require_database(database)?;
    detail_with_database(database, role_id).await?;
    let params = query.normalize_page()?;
    let user_ids = role_user_ids(database, role_id).await?;
    if user_ids.is_empty() {
        return Ok(PageResult::empty(params));
    }

    let offset = ((params.page - 1) * params.page_size) as i64;
    let placeholders = placeholders(user_ids.len());
    let mut args: Vec<rbs::Value> = user_ids.iter().map(|id| rbs::value!(id.value())).collect();
    args.push(rbs::value!(params.page_size));
    args.push(rbs::value!(offset));
    let rows: Vec<SysUser> = database
        .exec_decode(
            &format!(
                "select id, username, password_hash, password_salt, real_name, phone, email, avatar, status, dept_id, deleted, created_at, updated_at from sys_user where deleted = 0 and id in ({placeholders}) order by created_at desc limit ? offset ?"
            ),
            args,
        )
        .await
        .map_err(map_database_error)?;
    let total = count(
        database,
        &format!(
            "select count(*) as total from sys_user where deleted = 0 and id in ({placeholders})"
        ),
        user_ids.iter().map(|id| rbs::value!(id.value())).collect(),
    )
    .await?;
    Ok(PageResult {
        list: rows,
        total: I64String(total),
        page: params.page,
        page_size: params.page_size,
    })
}

pub async fn permissions(
    database: Option<&Arc<RBatis>>,
    role_id: i64,
) -> Result<RolePermissionVo, AppError> {
    let database = require_database(database)?;
    detail_with_database(database, role_id).await?;
    let rows: Vec<IdRow> = database
        .exec_decode(
            "select menu_id as id from sys_role_menu where role_id = ? order by menu_id asc",
            vec![rbs::value!(role_id)],
        )
        .await
        .map_err(map_database_error)?;
    Ok(RolePermissionVo {
        menu_ids: rows.into_iter().map(|row| row.id).collect(),
        permission_codes: Vec::new(),
    })
}

pub async fn create(
    database: Option<&Arc<RBatis>>,
    request: RoleSaveRequest,
) -> Result<I64String, AppError> {
    let database = require_database(database)?;
    validate_request(&request)?;
    let id = next_id();
    database
        .exec(
            "insert into sys_role (id, role_name, role_code, builtin, status, remark, deleted) values (?, ?, ?, 0, ?, ?, 0)",
            vec![
                rbs::value!(id),
                rbs::value!(request.role_name),
                rbs::value!(request.role_code),
                rbs::value!(request.status.unwrap_or(1)),
                rbs::value!(request.remark),
            ],
        )
        .await
        .map_err(map_database_error)?;
    Ok(I64String(id))
}

pub async fn update(
    database: Option<&Arc<RBatis>>,
    id: i64,
    request: RoleSaveRequest,
) -> Result<(), AppError> {
    let database = require_database(database)?;
    detail_with_database(database, id).await?;
    validate_request(&request)?;
    database
        .exec(
            "update sys_role set role_name = ?, role_code = ?, status = ?, remark = ? where id = ? and deleted = 0",
            vec![
                rbs::value!(request.role_name),
                rbs::value!(request.role_code),
                rbs::value!(request.status.unwrap_or(1)),
                rbs::value!(request.remark),
                rbs::value!(id),
            ],
        )
        .await
        .map_err(map_database_error)?;
    Ok(())
}

pub async fn delete(database: Option<&Arc<RBatis>>, id: i64) -> Result<(), AppError> {
    let database = require_database(database)?;
    let role = detail_with_database(database, id).await?;
    if role.builtin == 1 {
        return Err(AppError::bad_request("operation failed"));
    }
    let user_count = count(
        database,
        "select count(*) as total from sys_user_role where role_id = ?",
        vec![rbs::value!(id)],
    )
    .await?;
    if user_count > 0 {
        return Err(AppError::conflict("operation failed"));
    }
    database
        .exec(
            "update sys_role set deleted = 1 where id = ? and deleted = 0",
            vec![rbs::value!(id)],
        )
        .await
        .map_err(map_database_error)?;
    Ok(())
}

pub async fn update_status(
    database: Option<&Arc<RBatis>>,
    id: i64,
    status: i32,
) -> Result<(), AppError> {
    let database = require_database(database)?;
    detail_with_database(database, id).await?;
    database
        .exec(
            "update sys_role set status = ? where id = ? and deleted = 0",
            vec![rbs::value!(status), rbs::value!(id)],
        )
        .await
        .map_err(map_database_error)?;
    Ok(())
}

pub async fn assign_permissions(
    database: Option<&Arc<RBatis>>,
    role_id: i64,
    menu_ids: Vec<i64>,
) -> Result<(), AppError> {
    let database = require_database(database)?;
    detail_with_database(database, role_id).await?;
    assert_existing_menus(database, &menu_ids).await?;
    database
        .exec(
            "delete from sys_role_menu where role_id = ?",
            vec![rbs::value!(role_id)],
        )
        .await
        .map_err(map_database_error)?;
    for menu_id in menu_ids {
        database
            .exec(
                "insert into sys_role_menu (id, role_id, menu_id) values (?, ?, ?)",
                vec![
                    rbs::value!(next_id()),
                    rbs::value!(role_id),
                    rbs::value!(menu_id),
                ],
            )
            .await
            .map_err(map_database_error)?;
    }
    Ok(())
}

async fn detail_with_database(database: &RBatis, id: i64) -> Result<SysRole, AppError> {
    let rows: Vec<SysRole> = database
        .exec_decode(
            "select id, role_name, role_code, builtin, status, remark, deleted, created_at, updated_at from sys_role where id = ? and deleted = 0",
            vec![rbs::value!(id)],
        )
        .await
        .map_err(map_database_error)?;
    rows.into_iter()
        .next()
        .ok_or_else(|| AppError::not_found("资源不存在"))
}

async fn role_user_ids(database: &RBatis, role_id: i64) -> Result<Vec<I64String>, AppError> {
    let rows: Vec<IdRow> = database
        .exec_decode(
            "select user_id as id from sys_user_role where role_id = ? order by user_id asc",
            vec![rbs::value!(role_id)],
        )
        .await
        .map_err(map_database_error)?;
    Ok(rows.into_iter().map(|row| row.id).collect())
}

async fn assert_existing_menus(database: &RBatis, menu_ids: &[i64]) -> Result<(), AppError> {
    if menu_ids.is_empty() {
        return Ok(());
    }
    let unique_ids: BTreeSet<i64> = menu_ids.iter().copied().collect();
    if unique_ids.len() != menu_ids.len() {
        return Err(AppError::bad_request("operation failed"));
    }
    let total = count(
        database,
        &format!(
            "select count(*) as total from sys_menu where deleted = 0 and id in ({})",
            placeholders(unique_ids.len())
        ),
        unique_ids.iter().map(|id| rbs::value!(*id)).collect(),
    )
    .await?;
    if total != unique_ids.len() as i64 {
        return Err(AppError::bad_request("operation failed"));
    }
    Ok(())
}

fn role_where(query: &RoleQuery) -> (String, Vec<rbs::Value>) {
    let mut where_sql = String::from("where deleted = 0");
    let mut args = Vec::new();
    if let Some(role_name) = non_blank(query.role_name.as_deref()) {
        where_sql.push_str(" and role_name like ?");
        args.push(rbs::value!(format!("%{role_name}%")));
    }
    if let Some(role_code) = non_blank(query.role_code.as_deref()) {
        where_sql.push_str(" and role_code like ?");
        args.push(rbs::value!(format!("%{role_code}%")));
    }
    if let Some(status) = query.status {
        where_sql.push_str(" and status = ?");
        args.push(rbs::value!(status));
    }
    if let Some(created_at) = non_blank(query.created_at.as_deref()) {
        where_sql.push_str(" and created_at like ?");
        args.push(rbs::value!(format!("%{created_at}%")));
    }
    (where_sql, args)
}

async fn count(database: &RBatis, sql: &str, args: Vec<rbs::Value>) -> Result<i64, AppError> {
    let counts: Vec<CountRow> = database
        .exec_decode(sql, args)
        .await
        .map_err(map_database_error)?;
    Ok(counts.first().map(|row| row.total).unwrap_or(0))
}

fn require_database(database: Option<&Arc<RBatis>>) -> Result<&RBatis, AppError> {
    database
        .map(|database| database.as_ref())
        .ok_or_else(|| AppError::system("Rbatis database is not configured"))
}

fn validate_request(request: &RoleSaveRequest) -> Result<(), AppError> {
    require_text(&request.role_name, "roleName")?;
    require_text(&request.role_code, "roleCode")?;
    validate_length(&request.role_name, "roleName", 64)?;
    if !valid_role_code(&request.role_code) {
        return Err(AppError::bad_request("roleCode format is invalid"));
    }
    Ok(())
}

fn require_text(value: &str, field: &str) -> Result<(), AppError> {
    if value.trim().is_empty() {
        return Err(AppError::bad_request(format!("{field} is required")));
    }
    Ok(())
}

fn validate_length(value: &str, field: &str, max: usize) -> Result<(), AppError> {
    if value.chars().count() > max {
        return Err(AppError::bad_request(format!(
            "{field} length must be <= {max}"
        )));
    }
    Ok(())
}

fn valid_role_code(value: &str) -> bool {
    let mut chars = value.chars();
    matches!(chars.next(), Some(first) if first.is_ascii_uppercase())
        && value.chars().count() <= 64
        && value.chars().count() >= 2
        && chars.all(|ch| ch.is_ascii_uppercase() || ch.is_ascii_digit() || ch == '_')
}

fn placeholders(len: usize) -> String {
    vec!["?"; len].join(",")
}

fn non_blank(value: Option<&str>) -> Option<&str> {
    value.map(str::trim).filter(|value| !value.is_empty())
}

fn map_database_error(err: rbatis::Error) -> AppError {
    let message = err.to_string();
    if message.to_ascii_lowercase().contains("duplicate") {
        AppError::conflict("数据冲突")
    } else {
        AppError::system(message)
    }
}
